CIP-?/README.md

```diff
@@ -0,0 +1,718 @@
+only constant overheads. There are several examples of these, and all rely on
+bitwise operations for their implementations.
+
+For example, consider wanting to store a set of `BuiltinInteger`s
+on-chain. Given current on-chain primitives, the most viable option involves
+some variant on a `BuiltinList` of `BuiltinInteger`s; however,
+this is unviable in practice unless the set is small. To see why, suppose that
+we have an upper limit of $k$ on the `BuiltinInteger`s we want to store;
+this is realistic in practically all cases. To store $n$
+`BuiltinInteger`s under the above scheme requires 
+
+$$n \cdot \left( \left\lceil \frac{\log_2(k)}{64} \right\rceil \cdot 64  + c\right)
+$$
+
+bits, where $c$ denotes the constant overhead for each cons cell of
+the `BuiltinList` holding the data. If the set being represented is dense
+(meaning that the number of entries is a sizeable fraction of $k$), this cost
+becomes intolerable quickly, especially when taking into account the need to
+also store the operations manipulating such a structure on-chain with the script
+where the set is being used.
+
+If we instead represented the same set as a bitmap based on
+`BuiltinByteString`, the amount of space required would instead be 
+
+$$\left\lceil \frac{k}{8} \right\rceil \cdot 8 + \left\lceil
+\frac{\log_2(k)}{64} \right\rceil \cdot 64
+$$
+
+bits. This is significantly better unless $n$ is small. Furthermore,
+this representation would likely be more efficient in terms of time in practice,
+as instead of having to crawl through a cons-like structure, we can implement
+set operations on a memory-contiguous byte string:
+
+- The cardinality of the set can be computed as a population count. This
+can have terrifyingly efficient implementations: the Muła-Kurz-Lemire
+algorithm (the current state of the art) can process four kilobytes per loop
+iteration, which amounts to over four thousand potential stored integers.
+- Insertion or removal is a bit set or bit clear respectively.
+- Finding the smallest element is a find-first-one.
+- Testing for membership is a check to see if the bit is set.
+- Set intersection is bitwise and.
+- Set union is bitwise inclusive or.
+- Set symmetric difference is bitwise exclusive or.
+
+
+A potential implementation could use a range of techniques to make these
+operations extremely efficient, by relying on SWAR (SIMD-within-a-register)
+techniques if portability is desired, and SIMD instructions for maximum speed.
+This would allow both potentially large integer sets to be represented on-chain
+without breaking the size limit, and nodes to efficiently compute with such,
+reducing the usage of resources by the chain. Lastly, in practice, if
+compression techniques are used (which also rely on bitwise operations!), the
+number of required bits can be reduced considerably in most cases without
+compromising performance: the current state-of-the-art (Roaring Bitmaps) can be
+used as an example of the possible gains.
```

Comment:
This is definitely a consideration. It would depend heavily how much such a compressed structure would be used: if it's only for a small number of operations on small structures, it's probably not worth it; but if the number of operations needed, or the structure, is large, it's probably worthwhile.

Realistically, even in an off-chain setting, compressed versus uncompressed structures are a tradeoff which can't be automatically decided upon. Having bitwise primitives at least gives us the _option_ of making this choice!